Disable individual static rules in DNR (#33733)
mdn/content: files/en-us/mozilla/add-ons/webextensions/api/declarativenetrequest/updatestaticrules/index.md

page-type: webextension-api-function
browser-compat: webextensions.api.declarativeNetRequest.updateStaticRules
---

{{AddonSidebar}}

Modifies the enabled state of rules in a static ruleset.

Rules belonging to a ruleset can be modified while it is disabled but don't become effective until the ruleset is enabled.

## Syntax

```js-nolint
let updatedStaticRules = browser.declarativeNetRequest.updateStaticRules(
    options                // object
);
```

### Parameters

- `options`

  - : An object containing details of the ruleset and which of its rules are to be enabled or disabled.
    - `rulesetId` rulesetId
      - : `string` The ID of the static ruleset to modify.
    - `disableRuleIds` {{optional_inline}}
      - : An array of `number`. IDs of the rules to disable. Any invalid IDs are ignored.
    - `enableRuleIds` {{optional_inline}}
      - : An array of `number`. IDs of the rules to enable. Any invalid IDs are ignored.

### Return value

A [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) If the request is successful, the promise is fulfilled without arguments. If the request fails, the promise is rejected with an error message.

> {{WebExtExamples("h2")}}

Reviewer: Thought: if the example subsection is empty, wouldn't make sense to omit it for now and eventually add it if we will have an example to include in it?

Author: @rpl Nothing is rendered from this macro until the API is present in the [example.json file](https://github.com/mdn/webextensions-examples/blob/main/examples.json). Then, this macro adds the Examples title (as h2) and a list of examples (using the API).

Reviewer: @rebloor ah, I see, then I definitely agree that it would be better to include it as you did and leave the macro logic to eventually show the section and list of examples if we will have added one to the example.json file.

---
title: declarativeNetRequest.updateStaticRules
slug: Mozilla/Add-ons/WebExtensions/API/declarativeNetRequest/updateStaticRules
page-type: webextension-api-function
browser-compat: webextensions.api.declarativeNetRequest.updateStaticRules
---

{{AddonSidebar}}

Modifies the enabled state of rules in a static ruleset.

> Rules belonging to a ruleset can be modified while it is disabled but don't become effective until the ruleset is enabled.

Reviewer: Thought: how about "Rules belonging to a disabled ruleset can be modified and become effective when the ruleset is enabled"?

Author: Or should it be:
```suggestion
Rules in a ruleset can be modified while the ruleset or rules are disabled. Any changes become effective when the ruleset and rules are enabled.
```

Reviewer: @rebloor I'd just cut out the "or rules" and the "and rules" parts, this method is one that is menat to enable/disable rules and so I think the note can just focus on the fact that changing rules enabled in a disabled rules become effective when the ruleset may become enabled (on the contrary change to the enable/disable state of rules part of an enabled ruleset are immediately effective.

Author: ```suggestion
Rules can be enabled and disabled while the ruleset containing them is disabled. Any changes become effective when the ruleset is enabled.
```

Reviewer: Thanks Richard, the wording tweaks you mentioned in https://github.com/mdn/content/pull/33733#discussion_r1627480918 looks great to me too.

## Syntax

```js-nolint
let updatedStaticRules = browser.declarativeNetRequest.updateStaticRules(
    options                // object
);
```

### Parameters

- `options`

>   - : An object containing details of the ruleset and which of its rules are to be enabled or disabled.

Reviewer: Thought: I'm not sure the description for `options` is actually needed (and I wonder if the current wording may be a bit misleading), how about omitting it and just include description for its properties? alternatively we may call it "An object to specify which rules to enable/disable in a given static ruleset" or something along those lines.

Author: In general, we normally provide a description for an `options` objects. 
```suggestion
  - : An object specifying a static ruleset and the rules to enable or disable in that ruleset.
```

Reviewer: :+1:  I like the suggested version from your comment https://github.com/mdn/content/pull/33733#discussion_r1616558474

    - `rulesetId` rulesetId
      - : `string` The ID of the static ruleset to modify.
    - `disableRuleIds` {{optional_inline}}
      - : An array of `number`. IDs of the rules to disable. Any invalid IDs are ignored.
    - `enableRuleIds` {{optional_inline}}
      - : An array of `number`. IDs of the rules to enable. Any invalid IDs are ignored.

### Return value

A [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) If the request is successful, the promise is fulfilled without arguments. If the request fails, the promise is rejected with an error message.

> ## Examples
> 
> {{WebExtExamples}}

Author: ```suggestion
{{WebExtExamples("h2")}}
```
